blocks/search.py

```diff
@@ -0,0 +1,303 @@
+        states = []
+        for name in generator.generate.states:
+            var = VariableFilter(bricks=[generator], name='^' + name + '$',
+                                 roles=[INPUT])(inner_cg)[-1:]
+            if var:
+                self.need_input_states.append(name)
+            states.extend(var)
+        self.compile_context_computer(contexts)
+        self.compile_initial_state_computer(generator, contexts)
+        self.compile_next_state_computer(generator, contexts, states, inner_cg)
+        self.compile_costs_computer(generator, contexts, states, inner_cg)
+
+        self.compiled = True
+
+    def compute_contexts(self, inputs_dict):
+        """Computes contexts from inputs.
+
+        Wrapper around a theano function which precomputes contexts.
+
+        Parameters
+        ----------
+        inputs_dict : dict
+            Dictionary of input arrays.
+
+        """
+        contexts = self.context_computer(*[inputs_dict[name]
+                                           for name in self.inputs])
+        return OrderedDict(zip(self.context_names, contexts))
+
+    def compute_initial_states(self, contexts):
+        """Computes initial outputs and states.
+
+        Parameters
+        ----------
+        contexts : dict
+            Dictionary of contexts {name: context}.
+
+        """
+        init_states = self.initial_state_computer(*list(contexts.values()))
+        init_states = [state.reshape((1,) + state.shape)
+                       for state in init_states]
+        return OrderedDict(zip(self.state_names, init_states))
+
+    def compute_costs(self, contexts, cur_states):
+        """Computes next costs."""
+        states = [cur_states[name][0] for name in self.need_input_states]
+
+        logprobs = self.costs_computer(*(list(contexts.values()) + states))
+        return logprobs[None, :, :]
+
+    def compute_next_state(self, contexts, cur_states, outputs):
+        """Computes next states.
+
+        Parameters
+        ----------
+        contexts : dict
+            Dictionary of contexts.
+        cur_states : dict
+            Dictionary of current states.
+
+        Returns
+        -------
+        Dictionary of next state and probabilities values
+        with names as returned by `generate_outputs`.
+
+        """
+        # First timesteps only if state is needed
+        states = [cur_states[name][0] for name in self.need_input_states]
+
+        next_values = self.next_state_computer(*(list(contexts.values()) +
+                                                 states + [outputs]))
+
+        # Add time dimension back
+        next_values = [state.reshape((1,) + state.shape)
+                       for state in next_values]
+        return OrderedDict(zip(self.generate_names, next_values))
+
+    @staticmethod
+    def _top_probs(scores, beam_size, unique=False):
+        """Returns indexes of elements with lowest scores.
```

Comment:
You should do something with this function. It is very hard to understand why it receives 3D array of probabilities, to be honest I still do not get it after a week of staring at the code.